### ISSUE TITLE
Incorrect failure message attribute in junitxml file when combined with xdist

### ISSUE THREAD

**Issue by timyhou** [pytest-dev/pytest-xdist#130]
When combining xdist with pytest's junitxml flag there is an issue where the generated report will have the incorrect <failure message=> attribute value. 

test_case.py

```python
import pytest

class TestxDistXmlIssue:
    def test_one(self):
        # stack trace code
        # stack trace code
        # stack trace code
        # stack trace code
        # stack trace code
        assert 1 == 2, "Only print this message"

    def test_two(self):
        # stack trace code
        # stack trace code
        # stack trace code
        # stack trace code
        # stack trace code
        assert False, "Only print this message"
```

Ran with the following command

```
pytest -n2 --junitxml="test_xdist.xml" test_case.py
```

test_xdist.xml 

Notice that the <failure message=> attribute is a duplication of the stack trace 

```xml
<?xml version="1.0" encoding="utf-8"?>
<testsuite errors="0" failures="2" name="pytest" skips="0" tests="2" time="0.283">
    <testcase classname="pytesting.test_case.TestxDistXmlIssue" file="pytesting/test_case.py" line="25" name="test_two" time="0.0014913082122802734">
        <failure message="self = &lt;test_case.TestxDistXmlIssue object at 0x7f3770d5fa90&gt;      def test_two(self):         # stack trace code         # stack trace code         # stack trace code         # stack trace code         # stack trace code &gt;       assert False, &quot;Only print this message&quot; E       AssertionError: Only print this message E       assert False  pytesting/test_case.py:32: AssertionError">self = &lt;test_case.TestxDistXmlIssue object at 0x7f3770d5fa90&gt;

    def test_two(self):
        # stack trace code
        # stack trace code
        # stack trace code
        # stack trace code
        # stack trace code
&gt;       assert False, &quot;Only print this message&quot;
E       AssertionError: Only print this message
E       assert False

pytesting/test_case.py:32: AssertionError</failure>
    </testcase>
    <testcase classname="pytesting.test_case.TestxDistXmlIssue" file="pytesting/test_case.py" line="17" name="test_one" time="0.0014431476593017578">
        <failure message="self = &lt;test_case.TestxDistXmlIssue object at 0x7fab8351fac8&gt;      def test_one(self):         # stack trace code         # stack trace code         # stack trace code         # stack trace code         # stack trace code &gt;       assert 1 == 2, &quot;Only print this message&quot; E       AssertionError: Only print this message E       assert 1 == 2  pytesting/test_case.py:24: AssertionError">self = &lt;test_case.TestxDistXmlIssue object at 0x7fab8351fac8&gt;

    def test_one(self):
        # stack trace code
        # stack trace code
        # stack trace code
        # stack trace code
        # stack trace code
&gt;       assert 1 == 2, &quot;Only print this message&quot;
E       AssertionError: Only print this message
E       assert 1 == 2

pytesting/test_case.py:24: AssertionError</failure>
    </testcase>
</testsuite>
```

If we take off the -n2 flag to disable xdist we get the expected faliure message attribute in the report

test_no_xdist.xml

```xml
<?xml version="1.0" encoding="utf-8"?>
<testsuite errors="0" failures="2" name="pytest" skips="0" tests="2" time="0.024">
    <testcase classname="pytesting.test_case.TestxDistXmlIssue" file="pytesting/test_case.py" line="17" name="test_one" time="0.0010428428649902344">
        <failure message="AssertionError: Only print this message assert 1 == 2">self = &lt;test_case.TestxDistXmlIssue object at 0x7f63ae6656d8&gt;

    def test_one(self):
        # stack trace code
        # stack trace code
        # stack trace code
        # stack trace code
        # stack trace code
&gt;       assert 1 == 2, &quot;Only print this message&quot;
E       AssertionError: Only print this message
E       assert 1 == 2

pytesting/test_case.py:24: AssertionError</failure>
    </testcase>
    <testcase classname="pytesting.test_case.TestxDistXmlIssue" file="pytesting/test_case.py" line="25" name="test_two" time="0.00023293495178222656">
        <failure message="AssertionError: Only print this message assert False">self = &lt;test_case.TestxDistXmlIssue object at 0x7f63ae665358&gt;

    def test_two(self):
        # stack trace code
        # stack trace code
        # stack trace code
        # stack trace code
        # stack trace code
&gt;       assert False, &quot;Only print this message&quot;
E       AssertionError: Only print this message
E       assert False

pytesting/test_case.py:32: AssertionError</failure>
    </testcase>
</testsuite>
```

I believe the culprit is related to append_failure (line 133) in junitxml.py. When xdist is active that message is not being set to the expected longrepr.reprcrash.message value.

Additional information
ubuntu 14.04 python 3.6.0
Pip list
apipkg (1.4)
appdirs (1.4.3)
execnet (1.4.1)
packaging (16.8)
pip (9.0.1)
py (1.4.32)
pyparsing (2.2.0)
pytest (3.0.6)
pytest-xdist (1.15.0)
setuptools (34.3.1)
six (1.10.0)

**Comment by timyhou** [comment]
Any update on this? I am not very familiar with the xdist code, but I will try taking a look myself. Any hints would help greatly!

**Comment by nicoddemus** [comment]
Hi @timyhou, sorry for the silence on this, I think nobody got some free time to take a deeper look at it yet.

I suspect the problem is in the code which marshals the report in the worker node and sends it to the master node in https://github.com/pytest-dev/pytest-xdist/blob/master/xdist/remote.py#L100, if you want to take a look

**Comment by timyhou** [comment]
I think I was able to write a test which fits the same scenario as above in the pytest-xdist tests

```
class TestReportSerialization:
    def test_xdist_report_longrepr_reprcrash_130(self, testdir):
        reprec = testdir.inline_runsource("""
                    import py                    
                    def test_fail(): assert False, 'Expected Message'                    
                """, '-n1') # removing '-n1' will pass 
        reports = reprec.getreports("pytest_runtest_logreport")
        assert len(reports) == 3
        failure_report = reports[1]
        assert hasattr(failure_report.longrepr, 'reprcrash')
```

I know what I want to be able to see is the value of 'report' before and after is passes through the serialize_report method, but I am a little lost stepping through the code for the remote.py.

**Comment by nicoddemus** [comment]
Let me give a quick overview on how things related to that report object work:

In `remote.py`, the worker receives a [`pytest_runtest_logreport`](https://github.com/pytest-dev/pytest-xdist/blob/master/xdist/remote.py#L88) message from the hook system, serializes it as a `dict` and sends it to the `master` node.

The `master` node receives that event in [`slave_testreport`](https://github.com/pytest-dev/pytest-xdist/blob/master/xdist/dsession.py#L657) with the serialized contents. Hmmm at this point I'm not sure at which point the `data` dict sent by the remote is unserialized into a `TestReport` object, which is what `slave_testreport` seems to receive, and unfortunately I can't look deeper into it at the moment. Hopefully from this short overview you can figure it out though.

**Comment by mszpala** [comment]
Hi,
since I'm interested in fixing this issue I've started to looking into code. It seems that mechanism for serializing and unserializng is loosing 'reprcrash' as it's shown on test above. Probably longrepr from TestReport should be serialized and unserialized as a ReprExceptionInfo to have all the needed information and then unserialized in slavemanage.py.

**Comment by timyhou** [comment]
@mszpala  is right what the underlying issue is. 

One question is should `slavemanage.py` unserialize the `report.longrepr` as `ReprExceptionInfo` or `ExceptionChainRepr`? Which is what the type is initially. 

I have a super super ugly looking version locally which seems to fix the behavior though! The ugly part was deconstructing `longrepr` into dicts for serialization in `remote.py` and then putting it back together on the otherside in `slavemanage.py`.

**Comment by timyhou** [comment]
After some cleanup I have it working and all the tests passing properly! Should I push a branch? There are probably lots of improvements still to be made which you guys would know better.

**Comment by RonnyPfannschmidt** [comment]
please open a pull request, we are happy to help you bring your working experiment up to shape

👍

**Comment by nicoddemus** [comment]
Fixed in #164
